src/Interactive/EditorFeatures/Core/Extensibility/Interactive/SendToInteractiveSubmissionProvider.cs

```diff
@@ -0,0 +1,103 @@
+        protected abstract bool CanParseSubmission(string code);
+
+        public string GetSelectedText(IEditorOptions editorOptions, CommandArgs args, CancellationToken cancellationToken)
+        {
+
+            IEnumerable<SnapshotSpan> selectedSpans = args.TextView.Selection.GetSnapshotSpansOnBuffer(args.SubjectBuffer).Where(ss => ss.Length > 0);
+
+            // If there is no selection select the current line.
+            if (!selectedSpans.Any())
+            {
+                selectedSpans = GetSelectedLine(args);
+            }
+
+            // Send the selection as is if it does not contain any parsing errors.
+            var candidateSubmission = GetSubmissionFromSelectedSpans(editorOptions, selectedSpans);
+            if (CanParseSubmission(candidateSubmission))
+            {
+                return candidateSubmission;
+            }
+
+            // Otherwise heuristically try to expand it.
+            return GetSubmissionFromSelectedSpans(editorOptions, ExpandSelection(selectedSpans, args, cancellationToken));
+        }
+
+        /// <summary>Returns the span for the currently selected line.</summary>
+        private static IEnumerable<SnapshotSpan> GetSelectedLine(CommandArgs args)
+        {
+            SnapshotPoint? caret = args.TextView.GetCaretPoint(args.SubjectBuffer);
+            int caretPosition = args.TextView.Caret.Position.BufferPosition.Position;
+            ITextSnapshotLine containingLine = caret.Value.GetContainingLine();
+            return new SnapshotSpan[] {
+                new SnapshotSpan(containingLine.Start, containingLine.End)
+            };
+        }
+
+        private async Task<IEnumerable<SnapshotSpan>> GetExecutableSyntaxTreeNodeSelection(
+            TextSpan selectionSpan,
+            CommandArgs args,
+            ITextSnapshot snapshot,
+            CancellationToken cancellationToken)
+        {
+            Document doc = args.SubjectBuffer.GetRelatedDocuments().FirstOrDefault();
+            var semanticDocument = await SemanticDocument.CreateAsync(doc, cancellationToken).ConfigureAwait(false);
+            var text = semanticDocument.Text;
+            var root = semanticDocument.Root;
+            var model = semanticDocument.SemanticModel;
+
+            return GetExecutableSyntaxTreeNodeSelection(selectionSpan, text, root, model)
+                .Select(span => new SnapshotSpan(snapshot, span.Start, span.Length));
+        }
+
+        private IEnumerable<SnapshotSpan> ExpandSelection(IEnumerable<SnapshotSpan> selectedSpans, CommandArgs args, CancellationToken cancellationToken)
+        {
+            var selectedSpansStart = selectedSpans.Min(span => span.Start);
+            var selectedSpansEnd = selectedSpans.Max(span => span.End);
+            ITextSnapshot snapshot = args.TextView.TextSnapshot;
+
+            IEnumerable<SnapshotSpan> newSpans = GetExecutableSyntaxTreeNodeSelection(
+                TextSpan.FromBounds(selectedSpansStart, selectedSpansEnd),
```

Comment:
I think this makes sense. We want to expand the selection to cover a valid syntax node, which definitely can (must?) be covered by a stream selection that enclosing the box selection from user.